src/linquantarrays.jl

```diff
@@ -10,84 +13,164 @@ Base.size(QA::LinQuantArray) = size(QA.A)
 Base.getindex(QA::LinQuantArray,i...) = getindex(QA.A,i...)
 Base.eltype(Q::LinQuantArray{T,N}) where {T,N} = T
 
-"""Quantise an array linearly into a LinQuantArray."""
-function LinQuantization(::Type{T},A::AbstractArray) where {T<:Unsigned}
+"""
+
+        LinQuantization(::Type{T}, A::AbstractArray; extrema::Tuple = extrema(A)) where {T<:Integer}
+
+Quantise an array linearly into a LinQuantArray.
+
+# Arguments
+- `T`: the type of the quantised array
+- `A`: the array to quantise
+- `extrema`: the minimum and maximum of the range, defaults to `extrema(A)`.
+
+# Returns
+- a LinQuantArray{T} with the quantised array and the minimum and maximum of the original range.
+"""
+
+function LinQuantization(
+    ::Type{T},
+    A::AbstractArray,
+    extrema::Nothing
+) where {T<:Integer}
     all(isfinite.(A)) || throw(DomainError("Linear quantization only in (-∞,∞)"))
 
-    Amin = Float64(minimum(A))              # minimum of value range
-    Amax = Float64(maximum(A))              # maximum of value range
+    # range of values in A
+    Amin, Amax  = Float64(minimum(A)), Float64(maximum(A))    
 
-    if Amin == Amax
-        Δ = 0.0                                 # set to zero for no range
-    else
-        Δ = (2^(sizeof(T)*8)-1)/(Amax-Amin)     # inverse spacing
-    end
+    # minimum and maximum representable value of type T
+    Tmin, Tmax = Float64(typemin(T)), Float64(typemax(T))
+
+    # inverse spacing, set to zero for no range
+    Δ⁻¹ = Amin == Amax ? zero(Float64) : (Tmax-Tmin)/(Amax-Amin)
 
     Q = similar(A,T)                        # preallocate
 
-    # map minimum to 0x0, maximum to 0xff...ff
+    # map minimum to typemin(T), maximum to typemax(t)
     @inbounds for i in eachindex(Q)
-        Q[i] = round((A[i]-Amin)*Δ)
+        Q[i] = round((A[i]-Amin)*Δ⁻¹ + Tmin)
     end
 
     return LinQuantArray{T,ndims(Q)}(Q,Amin,Amax)
 end
 
-# define 8, 16, 24 and 32 bit
+function LinQuantization(
+    ::Type{T},
+    A::AbstractArray,
+    extrema::Tuple = extrema(A),
```

Comment:
I think I'd prefer to turn this into a keyword argument?

Reply:
Multiple dispatch doesn't support keyword arguments. If there is:

```julia 
function LinQuantization(
    ::Type{T},
    A::AbstractArray;
    extrema::Nothing=nothing
) where {T<:Integer}
...
end
```
and 
```julia 
function LinQuantization(
    ::Type{T},
    A::AbstractArray;
    extrema::Tuple
) where {T<:Integer}
...
end
```
the second definition will override the first one. To make it keyword argument, we could do the following: 

```julia 

function _LinQuantization(
    ::Type{T},
    A::AbstractArray;
    extrema::Union{Tuple, Nothing} = nothing
) where {T<:Integer}
  isnothing(extrema) ? _LinQuantization(T,A) : _LinQuantization(T,A,extrema)
end

function _LinQuantization(
    ::Type{T},
    A::AbstractArray
) where {T<:Integer}
...
end

function _LinQuantization(
    ::Type{T},
    A::AbstractArray
    extrema::Tuple
) where {T<:Integer}
...
end
```